source/on.js

```diff
@@ -0,0 +1,31 @@
+import curry from './curry';
+
+
+/**
+ * Takes a binary curried function `f`, a unary function `g`, and two values.
+ * Applies `g` to each value, then applies the result of each to `f`.
+ *
+ * Also known as the P combinator.
+ *
+ * @func
+ * @memberOf R
+ * @category Function
+ * @sig f -> g -> a -> b -> c
```

Comment:
I believe this should read:

     * @sig ((b, b) => c) -> (a -> b) -> a -> a -> c

Since `on` is a curried form of (f, g, x, y) => f ( g (x), g (y) ), we can map `a` as the type of `x` and `y`, `b` as the type of `g (x)` and of `g (y)`, and then `g` has the type `a  -> b`, and `f` has the type `(b, b) -> c`  Another way to think about it is that we have three fundamental types, the input type (`a) for which we'll supply to a function (`f`) to convert two values into an output type (`c`) by way of a conversion function (`g`) that generates an intermediate type (`b`) and is applied to each input.

Wadler's famous paper, [Theorems for Free][pw] should demonstrate that this signature strongly restricts the number of possible  implementations that can match.  If I'm not mistaken, after currying it could only be one of these:

```js
(f) => (g) => (x) => (y) => f ( g (x), g (x) )
(f) => (g) => (x) => (y) => f ( g (x), g (y) )
(f) => (g) => (x) => (y) => f ( g (y), g (x) )
(f) => (g) => (x) => (y) => f ( g (y), g (y) )
```

 ... and that is one advantage of such type signatures.


  [pw]: https://people.mpi-sws.org/~dreyer/tor/papers/wadler.pdf